pkg/serviceprovider/quay/metadataprovider.go

```diff
@@ -0,0 +1,111 @@
+//
+// Copyright (c) 2021 Red Hat, Inc.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package quay
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+	"net/url"
+
+	api "github.com/redhat-appstudio/service-provider-integration-operator/api/v1beta1"
+	"github.com/redhat-appstudio/service-provider-integration-operator/pkg/serviceprovider"
+	"github.com/redhat-appstudio/service-provider-integration-operator/pkg/spi-shared/tokenstorage"
+)
+
+type metadataProvider struct {
+	httpClient   *http.Client
+	tokenStorage tokenstorage.TokenStorage
+	quay         Quay
+}
+
+var _ serviceprovider.MetadataProvider = (*metadataProvider)(nil)
+var quayUserApiEndpoint *url.URL
+
+func init() {
+	qUrl, err := url.Parse("https://quay.io/api/v1/user")
+	if err != nil {
+		panic(err)
+	}
+	quayUserApiEndpoint = qUrl
+}
+
+func (s metadataProvider) Fetch(ctx context.Context, token *api.SPIAccessToken) (*api.TokenMetadata, error) {
+	data, err := s.tokenStorage.Get(ctx, token)
+	if err != nil {
+		return nil, err
+	}
+
+	if data == nil {
+		return nil, err
+	}
+
+	state := &TokenState{
+		AccessibleRepos: map[RepositoryUrl]RepositoryRecord{},
+	}
+	//TODO: verify access
+
+	username, err := s.fetchUser(data.AccessToken)
+	if err != nil {
+		return nil, err
+	}
+
+	state.RemoteUsername = username
+
+	js, err := json.Marshal(state)
+	if err != nil {
+		return nil, err
+	}
+
+	metadata := token.Status.TokenMetadata
+	if metadata == nil {
+		metadata = &api.TokenMetadata{}
+		token.Status.TokenMetadata = metadata
+	}
+
+	scopes := make([]string, 0)
+	for _, p := range token.Spec.Permissions.Required {
```

Comment:
You're missing `token.Spec.Permissions.AdditionalScopes` here.

There is a utility method in the `serviceprovider` package for this kind of thing:

Extract `quay.TranslateToScopes` into a simple function (not a struct method), call it `translateToQuayScopes` or something and use `serviceprovider.GetAllScopes(translateToQuayScopes, &token.Spec.Permissions)` instead of the loop here.

Reply:
ok, applied